File: backend/src/main/java/com/mannschaft/app/shift/ShiftPreference.java
package com.mannschaft.app.shift;

/**
 * シフト希望の優先度区分（F03.5 v2: 5 段階）。
 *
 * <p>並び順は優先度の高（出勤したい）から低（絶対休み）へ。自動割当時のスコア計算で
 * この順序（ordinal）を参照する場合があるため、定義順の変更は慎重に行うこと。</p>
 *
 * <p>v1 の 3 段階（PREFERRED / AVAILABLE / UNAVAILABLE）は v2 で 5 段階に拡張された。
 * 旧 {@code UNAVAILABLE} は {@link #STRONG_REST} に移行済み（Flyway V3.137）。</p>
 */
public enum ShiftPreference {

    /** 希望（出勤したい）— 自動割当スコア: +100（設計書 F03.5 v2.3 §5.10 準拠）。 */
    PREFERRED,

    /** 可能（指定なし・出勤可能）— 自動割当スコア: 0。 */
    AVAILABLE,

    /** できれば出勤を避けたい（出れなくはない）— 自動割当スコア: -30（設計書 F03.5 v2.3 §5.10 準拠）。 */
    WEAK_REST,

    /** できれば休みたい（旧 UNAVAILABLE 相当）— 自動割当スコア: -80。 */
    STRONG_REST,

    /** 絶対休み（自動割当の対象外とする強い拒否）— 自動割当スコア: -∞（候補から除外）。 */
    ABSOLUTE_REST;

    /**
     * 自動割当の候補除外フラグ。{@link #ABSOLUTE_REST} のみ候補から完全除外される。
     *
     * @return {@code true} = 候補から完全除外 / {@code false} = スコア計算対象
     */
    public boolean isHardExcluded() {
        return this == ABSOLUTE_REST;
    }

    /** {@link #isHardExcluded()} の逆。割当可能かどうかを返す（互換用）。 */
    public boolean isAssignable() {
        return !isHardExcluded();
    }

    /**
     * 自動割当スコアリング用の評価点。値が大きいほどアサイン優先度が高い。
     *
     * <p>PREFERRED=+100 / AVAILABLE=0 / WEAK_REST=-30 / STRONG_REST=-80 / ABSOLUTE_REST=-∞。
     * 設計書 F03.5 v2.3 §5.10 準拠。</p>
     *
     * @return スコア値
     */
    public int toAssignmentScore() {
        return switch (this) {
            case PREFERRED -> 100;
            case AVAILABLE -> 0;
            case WEAK_REST -> -30;
            case STRONG_REST -> -80;
            case ABSOLUTE_REST -> Integer.MIN_VALUE;
        };
    }
}

File: frontend/app/utils/shiftPreference.ts
import type { ShiftPreference } from '~/types/shift'

/**
 * F03.5 シフト希望強度ユーティリティ。
 *
 * 5段階の ShiftPreference に対して i18n キー・Tailwind 色クラス・割当スコアを提供する。
 * 設計書§5 のラジオカード 5色仕様に準拠:
 *   ABSOLUTE_REST=赤, STRONG_REST=橙, WEAK_REST=黄, AVAILABLE=灰, PREFERRED=緑
 */

/**
 * ShiftPreference を i18n キーに変換する。
 * @param preference 希望強度
 * @returns i18n キー文字列（例: `'shift.preference.preferred'`）
 */
export function preferenceToI18nKey(preference: ShiftPreference): string {
  const map: Record<ShiftPreference, string> = {
    PREFERRED: 'shift.preference.preferred',
    AVAILABLE: 'shift.preference.available',
    WEAK_REST: 'shift.preference.weakRest',
    STRONG_REST: 'shift.preference.strongRest',
    ABSOLUTE_REST: 'shift.preference.absoluteRest',
  }
  return map[preference]
}

/**
 * ShiftPreference を Tailwind 背景色クラスに変換する。
 * ラジオカード UI（SelectButton）のチップ色として使用する。
 * @param preference 希望強度
 * @returns Tailwind CSS クラス文字列
 */
export function preferenceToColor(preference: ShiftPreference): string {
  const map: Record<ShiftPreference, string> = {
    PREFERRED: 'bg-green-100 text-green-800 border-green-300',
    AVAILABLE: 'bg-gray-100 text-gray-700 border-gray-300',
    WEAK_REST: 'bg-yellow-100 text-yellow-800 border-yellow-300',
    STRONG_REST: 'bg-orange-100 text-orange-800 border-orange-300',
    ABSOLUTE_REST: 'bg-red-100 text-red-800 border-red-300',
  }
  return map[preference]
}

/**
 * ShiftPreference を自動割当スコアに変換する。
 *
 * スコア定義（設計書 F03.5 v2.3 §5.10 準拠・Phase 1 MVP 実装値）:
 * | 値             | スコア            |
 * |---|---|
 * | PREFERRED      | +100 （出勤希望）  |
 * | AVAILABLE      | 0    （指定なし）   |
 * | WEAK_REST      | -30  （出れなくはない） |
 * | STRONG_REST    | -80  （できれば休み）   |
 * | ABSOLUTE_REST  | -Infinity（割当不可）  |
 *
 * 旧 v2.2 の +50/-20 から Phase 1 MVP で +100/-30 へ上方修正し、設計書側を v2.3 として追随。
 * バックエンドの {@code ShiftPreference#toAssignmentScore()} と一致。Phase 2 の自動割当実機
 * テスト結果を踏まえ、動的重み調整の余地を残す。
 * @param preference 希望強度
 * @returns 割当スコア（ABSOLUTE_REST は Number.NEGATIVE_INFINITY）
 */
export function preferenceToScore(preference: ShiftPreference): number {
  const map: Record<ShiftPreference, number> = {
    PREFERRED: 100,
    AVAILABLE: 0,
    WEAK_REST: -30,
    STRONG_REST: -80,
    ABSOLUTE_REST: Number.NEGATIVE_INFINITY,
  }
  return map[preference]
}
